File: static/js/admin-v3-theme.js
/*
 * admin-v3-theme.js — VayuPress Admin v3 Theme Studio.
 *
 * Vanilla JS, strict CSP: no eval, no innerHTML with untrusted data, no inline
 * <style> injection. The live preview is driven entirely through the CSSOM
 * (element.style.setProperty for --vp-* variables) — scripted style writes are
 * not gated by style-src, so no compiled-CSS string is ever parsed client side.
 *
 * Flow:
 *   1. Load active tokens (/os/api/theme/tokens) → fill inputs + preview.
 *   2. Load presets (/os/api/theme/presets) → gallery of swatch cards.
 *   3. Editing any token live-updates the preview locally.
 *   4. "Apply theme" POSTs the full token set to /os/api/theme/apply,
 *      which validates + compiles + persists server-side (the only source of
 *      truth). "Revert" reloads the persisted tokens.
 */
(function () {
  'use strict';

  var root = document.querySelector('[data-theme-studio]');
  if (!root) return;

  var presetsEl = root.querySelector('[data-theme-presets]');
  var previewEl = root.querySelector('[data-theme-preview]');
  var statusEl = document.querySelector('[data-theme-status]');
  var applyBtn = document.querySelector('[data-theme-apply]');
  var revertBtn = document.querySelector('[data-theme-revert]');

  // All token inputs keyed by canonical field name.
  var inputs = {};
  root.querySelectorAll('[data-token]').forEach(function (el) {
    inputs[el.getAttribute('data-token')] = el;
  });

  // current token model (field name → value string).
  var model = {};

  function setStatus(msg, kind) {
    if (!statusEl) return;
    statusEl.textContent = msg;
    statusEl.className = 'text-sm' + (kind ? ' status--' + kind : ' muted');
  }

  function csrfToken() {
    var m = document.cookie.match(/(?:^|;\s*)vp_csrf=([^;]+)/);
    return m ? decodeURIComponent(m[1]) : '';
  }

  // Apply the model to the live preview by setting --vp-* custom properties on
  // the preview container (dark-mode tokens only — those carry data-token-var).
  function paintPreview() {
    root.querySelectorAll('[data-token-var]').forEach(function (el) {
      var field = el.getAttribute('data-token');
      var vari = el.getAttribute('data-token-var');
      var val = model[field];
      if (val) previewEl.style.setProperty('--vp-' + vari, val);
    });
    // Typography/layout, when valid CSS-ish, also drive the preview.
    if (model.FontSans) previewEl.style.setProperty('--vp-font-sans', model.FontSans);
    if (model.FontMono) previewEl.style.setProperty('--vp-font-mono', model.FontMono);
    if (model.FontSizeBase) previewEl.style.setProperty('--vp-font-size-base', model.FontSizeBase);
    if (model.LineHeight) previewEl.style.setProperty('--vp-line-height', model.LineHeight);
    if (model.RadiusLg) previewEl.style.setProperty('--vp-radius-lg', model.RadiusLg);
  }

  // Fill the form inputs and model from a token object (Go field names).
  function loadTokens(tok) {
    model = {};
    Object.keys(inputs).forEach(function (field) {
      var val = tok[field] != null ? String(tok[field]) : '';
      model[field] = val;
      var el = inputs[field];
      if (el.type === 'color') {
        // <input type=color> needs a 7-char hex; ignore shorthand/empty safely.
        if (/^#[0-9a-fA-F]{6}$/.test(val)) el.value = val;
      } else {
        el.value = val;
      }
    });
    paintPreview();
  }

  // Wire input → model → preview.
  Object.keys(inputs).forEach(function (field) {
    var el = inputs[field];
    var evt = el.type === 'color' ? 'input' : 'change';
    el.addEventListener(evt, function () {
      model[field] = el.value;
      paintPreview();
    });
  });

  // ── Presets gallery ─────────────────────────────────────────────────────────
  function swatch(color) {
    var s = document.createElement('span');
    s.className = 'theme-preset__swatch';
    if (/^#[0-9a-fA-F]{3,8}$/.test(color || '')) {
      s.style.setProperty('background-color', color);
    }
    return s;
  }

  function renderPresets(list) {
    while (presetsEl.firstChild) presetsEl.removeChild(presetsEl.firstChild);
    list.forEach(function (p) {
      var card = document.createElement('button');
      card.type = 'button';
      card.className = 'theme-preset';

      var swatches = document.createElement('div');
      swatches.className = 'theme-preset__swatches';
      [p.BgDark, p.SurfaceDark, p.AccentDark, p.Accent2Dark, p.HiDark].forEach(function (c) {
        swatches.appendChild(swatch(c));
      });

      var name = document.createElement('div');
      name.className = 'theme-preset__name';
      name.textContent = p.Name || 'Preset';

      card.appendChild(swatches);
      card.appendChild(name);
      card.addEventListener('click', function () {
        loadTokens(p);
        setStatus('Preset “' + (p.Name || '') + '” loaded — not yet applied', 'warn');
      });
      presetsEl.appendChild(card);
    });
  }

  // ── Persistence ─────────────────────────────────────────────────────────────
  function apply() {
    setStatus('Applying…');
    fetch('/os/api/theme/apply', {
      method: 'POST',
      headers: { 'Content-Type': 'application/json', 'X-CSRF-Token': csrfToken() },
      body: JSON.stringify({ tokens: model })
    }).then(function (r) {
      if (!r.ok) return r.json().then(function (e) { throw new Error((e.error && e.error.message) || ('apply failed (' + r.status + ')')); });
      return r.json();
    }).then(function (d) {
      setStatus('Applied · ' + (d.name || '') + ' · ' + new Date().toLocaleTimeString(), 'ok');
      if (window.vpToast) window.vpToast('Theme applied', 'ok');
    }).catch(function (err) {
      setStatus(String(err.message || err), 'danger');
    });
  }

  function fetchTokens() {
    return fetch('/os/api/theme/tokens', { headers: { Accept: 'application/json' } })
      .then(function (r) { return r.ok ? r.json() : Promise.reject(r.status); })
      .then(function (tok) { loadTokens(tok); });
  }

  function fetchPresets() {
    return fetch('/os/api/theme/presets', { headers: { Accept: 'application/json' } })
      .then(function (r) { return r.ok ? r.json() : Promise.reject(r.status); })
      .then(function (list) { renderPresets(Array.isArray(list) ? list : []); });
  }

  if (applyBtn) applyBtn.addEventListener('click', apply);
  if (revertBtn) revertBtn.addEventListener('click', function () {
    fetchTokens().then(function () { setStatus('Reverted to saved theme', 'ok'); });
  });

  // ── Init ────────────────────────────────────────────────────────────────────
  Promise.all([fetchTokens(), fetchPresets()])
    .then(function () { setStatus('Ready'); })
    .catch(function () { setStatus('Could not load theme', 'danger'); });
})();

File: cmd/vayupress/admin_v3_theme.go
package main

// admin_v3_theme.go — Admin v3 "Theme Studio" surface (VayuOS consolidation).
//
// Folds the v1/v2 Theme Studio (preset gallery + design-token editor + live
// preview) into the single v3 admin. The heavy lifting — preset definitions,
// hex/font/dimension validation, CSS compilation, persistence — already lives in
// internal/theme and the shared handlers (handleThemePresets/Tokens/Preview/
// Apply). This file adds the v3 page shell; the JSON endpoints are reused under
// session-friendly /os/api/theme/* mirrors registered in admin_v3_ui.go.
//
// CSP posture matches the rest of admin v3: zero inline styles, the only inline
// <script> carries the per-request nonce, every dynamic string is escaped. The
// live preview never injects a <style> element — it sets --vp-* custom
// properties on the preview container through the CSSOM (scripted style writes
// are not gated by style-src), so no compiled-CSS string is ever parsed client
// side.

import (
	"html"
	"net/http"

	"github.com/johalputt/vayupress/internal/render"
)

// themeColorField is one editable colour token. Field is the canonical Tokens
// field name (matched case-insensitively by applyOverrides); Vari is the public
// --vp-* variable it maps to for the live preview (dark-mode tokens only).
type themeColorField struct {
	Field string // e.g. "AccentDark"
	Label string
	Vari  string // e.g. "accent" → --vp-accent (empty when not previewed)
}

// themeDarkColors are the dark-mode tokens, each wired to a preview variable.
func themeDarkColors() []themeColorField {
	return []themeColorField{
		{"BgDark", "Background", "bg"},
		{"SurfaceDark", "Surface", "surface"},
		{"TextDark", "Text", "text"},
		{"MutedDark", "Muted", "muted"},
		{"AccentDark", "Accent", "accent"},
		{"Accent2Dark", "Accent 2", "accent2"},
		{"HiDark", "Highlight", "hi"},
		{"GreenDark", "Success", "green"},
	}
}

// themeLightColors are the light-mode tokens (no live preview — preview is dark).
func themeLightColors() []themeColorField {
	return []themeColorField{
		{"BgLight", "Background", ""},
		{"SurfaceLight", "Surface", ""},
		{"TextLight", "Text", ""},
		{"MutedLight", "Muted", ""},
		{"AccentLight", "Accent", ""},
		{"Accent2Light", "Accent 2", ""},
		{"HiLight", "Highlight", ""},
	}
}

// colorRow renders one colour-token control. The colour input carries the
// canonical field name and (when set) the preview variable so the JS can both
// serialise the token and live-update the preview without a server round-trip.
func colorRow(f themeColorField) string {
	vari := ""
	if f.Vari != "" {
		vari = ` data-token-var="` + html.EscapeString(f.Vari) + `"`
	}
	return `<label class="theme-field">
  <span class="theme-field__label">` + html.EscapeString(f.Label) + `</span>
  <input type="color" class="theme-field__color" data-token="` + html.EscapeString(f.Field) + `"` + vari + ` aria-label="` + html.EscapeString(f.Label) + `">
</label>`
}

// textRow renders a typography/layout text token control.
func textRow(field, label, placeholder string) string {
	return `<label class="theme-field theme-field--text">
  <span class="theme-field__label">` + html.EscapeString(label) + `</span>
  <input type="text" class="input" data-token="` + html.EscapeString(field) + `" placeholder="` + html.EscapeString(placeholder) + `" aria-label="` + html.EscapeString(label) + `">
</label>`
}

func (a *App) handleV3Theme(w http.ResponseWriter, r *http.Request) {
	nonce := render.CSPNonce(r)
	cfg := a.getV3Settings(r.Context())

	darkRows := ""
	for _, f := range themeDarkColors() {
		darkRows += colorRow(f)
	}
	lightRows := ""
	for _, f := range themeLightColors() {
		lightRows += colorRow(f)
	}
	typoRows := textRow("FontSans", "Sans-serif stack", "system-ui, sans-serif") +
		textRow("FontMono", "Monospace stack", "ui-monospace, monospace") +
		textRow("FontSizeBase", "Base font size", "1rem") +
		textRow("LineHeight", "Line height", "1.6") +
		textRow("MaxWidth", "Max content width", "72ch") +
		textRow("RadiusSm", "Small radius", "0.25rem") +
		textRow("RadiusLg", "Large radius", "0.75rem")

	body := `<div class="page-header">
  <h1>Theme Studio</h1>
  <div class="page-actions">
    <span class="text-sm muted" data-theme-status>Loading…</span>
    <button type="button" class="btn btn--ghost btn--sm" data-theme-revert>Revert</button>
    <button type="button" class="btn btn--primary btn--sm" data-theme-apply>Apply theme</button>
  </div>
</div>

<div class="theme-studio" data-theme-studio>
  <div class="theme-studio__main">
    <div class="card mb-6">
      <div class="card-title">Presets</div>
      <div class="text-sm muted mb-3">Start from a built-in palette, then fine-tune any token below.</div>
      <div class="theme-presets" data-theme-presets aria-label="Theme presets"></div>
    </div>

    <div class="card mb-6">
      <div class="card-title">Dark mode colours</div>
      <div class="theme-fields">` + darkRows + `</div>
    </div>

    <div class="card mb-6">
      <div class="card-title">Light mode colours</div>
      <div class="theme-fields">` + lightRows + `</div>
    </div>

    <div class="card">
      <div class="card-title">Typography &amp; layout</div>
      <div class="theme-fields theme-fields--text">` + typoRows + `</div>
    </div>
  </div>

  <aside class="theme-studio__preview" aria-label="Live preview">
    <div class="card-title">Live preview</div>
    <div class="theme-prev" data-theme-preview>
      <article class="theme-prev__article">
        <h1 class="theme-prev__h1">The quick brown fox</h1>
        <p class="theme-prev__p">VayuPress is a <strong>sovereign, observable</strong> publishing runtime. Every colour and typeface is a tunable token.</p>
        <h2 class="theme-prev__h2">Getting started</h2>
        <p class="theme-prev__p">Pick a preset, tweak a token, and click <em>Apply</em>. Changes reach <code class="theme-prev__code">/theme.css</code> immediately.</p>
        <blockquote class="theme-prev__quote">Simplicity is the ultimate sophistication.</blockquote>
        <p class="theme-prev__p"><a class="theme-prev__link" href="#">Read the documentation →</a></p>
        <div class="theme-prev__chips">
          <span class="theme-prev__chip theme-prev__chip--accent">accent</span>
          <span class="theme-prev__chip theme-prev__chip--hi">highlight</span>
          <span class="theme-prev__chip theme-prev__chip--green">success</span>
        </div>
      </article>
    </div>
    <div class="text-xs muted mt-3">Preview reflects dark-mode tokens. Light-mode values apply on readers whose system is set to light.</div>
  </aside>
</div>
<script nonce="` + nonce + `" src="/os/static/js/admin-v3-theme.js"></script>`

	writeV3HTML(w, adminV3Layout(nonce, "Theme Studio", "theme", cfg, body))
}
